src/node.js

```diff
@@ -390,6 +391,61 @@
     }
   };
 
+  startup.processPromises = function() {
+    // this should be kept in sync with PromiseRejectEvent in v8.h
+    var kPromiseRejectWithNoHandler = 0;
+    var kPromiseHandlerAddedAfterReject = 1;
+
+    var timers = NativeModule.require('timers');
+    var pendingRejections = [];
```

Comment:
@petkaantonov I'm not sure what that twit proves or shows. @vkurchatkin said he is worried about the impact of `.splice`ing for removing from the array.

At the moment [deleting from a map](https://github.com/v8/v8/blob/master/src/collection.js#L230) does [this](https://github.com/v8/v8/blob/2816532d00405d1adc1b546d1ab464277537ed86/src/runtime/runtime-collections.cc#L150-L161) - removing from an OrderedHashMap which in turn [inherits from OrderedHashTable](https://github.com/v8/v8-git-mirror/blob/master/src/objects.h#L4019) which has a [Remove](https://github.com/v8/v8-git-mirror/blob/master/src/objects.h#L3850-L3999) implemented in objects.cc line 16104 calling RemoveEntry in L16207 which puts a hole in the map where the entry was. The crux here is that ECMAScript maps are ordered by insertion order so they have to maintain order anyway - so there is something akin to an array splice going on anyway - even if v8 will improve its implementation.

For sizes this big it shouldn't matter very much anyway since rejections are rare. If we want performance that is at the theoretical bound of a map we can put `null`s in the array instead of splicing and skipping them during iteration - like a map does.